sympy/sets/fancysets.py

```diff
@@ -153,21 +155,38 @@ def _boundary(self):
         return self
 
     def _eval_imageset(self, f):
+        from sympy.simplify.simplify import clear_coefficients
+        from sympy.functions.elementary.integers import floor
+
         expr = f.expr
+        if not isinstance(expr, Expr):
+            return
+
         if len(f.variables) > 1:
             return
-        n = f.variables[0]
 
-        a = Wild('a')
-        b = Wild('b')
+        n = f.variables[0]
 
-        match = expr.match(a*n + b)
-        if match[a].is_negative:
-            expr = -expr
+        x, r = clear_coefficients(expr, 0)
+        if x == n:
+            # shift origin canonically
+            r = floor(r)
+            expr = expr.subs(n, n + r)
 
+        a = Wild('a', exclude=[n])
+        b = Wild('b', exclude=[n])
         match = expr.match(a*n + b)
-        if match[a] is S.One and match[b].is_integer:
-            expr = expr - match[b]
+        if match:
+            # make sign on x positive
+            if match[a].is_negative:
+                expr = -expr
```

Comment:
This appears to modify the image. We are expected to have the constant  b  in the image for  n = 0. After this change we seem to get  -b  instead. (Unless  b  and  -b  differ by a multiple of  a.)